src/index.ts

```diff
@@ -13,49 +13,80 @@
 
+// Returns value of metadata field.
+// Throws error if there is a problem accessing metadata API.
+async function getMetadataInstanceField(field: string): Promise<string> {
+  const [response, metadata] =
+      await pify(gcpMetadata.instance, {multiArgs: true})(field);
+  return metadata;
+}
+
 // initConfig sets unset values in the configuration to the value retrieved from
-// environment variables, metadata, or the default value specified in
+// environment variables, metadata, or the default values specified in
 // defaultConfig.
-// Returns rejected promise if value that must be set cannot be initialized.
-async function initConfig(config: Config): Promise<ProfilerConfig> {
+// Throws error if value that must be set cannot be initialized.
+// Exported for testing purposes.
+export async function initConfig(config: Config): Promise<ProfilerConfig> {
   config = common.util.normalizeArguments(null, config);
 
-  const envConfig = {
-    logLevel: process.env.GCLOUD_PROFILER_LOGLEVEL,
+  const envConfig: Config = {
     projectId: process.env.GCLOUD_PROJECT,
     serviceContext: {
       service: process.env.GAE_SERVICE,
       version: process.env.GAE_VERSION,
     }
   };
 
+  if (process.env.GCLOUD_PROFILER_LOGLEVEL !== undefined) {
+    let envLogLevel = parseInt(process.env.GCLOUD_PROFILER_LOGLEVEL || '', 10);
+    if (envLogLevel !== NaN) {
+      envConfig.logLevel = envLogLevel;
+    }
+  }
+
   let envSetConfig: Config = {};
   if (process.env.hasOwnProperty('GCLOUD_PROFILER_CONFIG')) {
     envSetConfig =
         require(path.resolve(process.env.GCLOUD_PROFILER_CONFIG)) as Config;
   }
 
-  let normalizedConfig = extend(true, {}, defaultConfig, envConfig, config);
+  let normalizedConfig =
+      extend(true, {}, defaultConfig, envSetConfig, envConfig, config);
 
-  if (normalizedConfig.serviceContext.service === undefined) {
-    throw new Error('service name must be specified in the configuration');
+  if (!normalizedConfig.zone || !normalizedConfig.instance) {
+    const [instance, zone] =
+        await Promise
+            .all([
+              getMetadataInstanceField('name'), getMetadataInstanceField('zone')
+            ])
+            .catch(
+                (err: Error) => {
+                    // ignore errors, which will occur when not on GCE.
+                }) ||
+        ['', ''];
+    if (!normalizedConfig.zone) {
+      normalizedConfig.zone = zone.substring(zone.lastIndexOf('/') + 1);
+    }
+    if (!normalizedConfig.instance) {
+      normalizedConfig.instance = instance;
+    }
   }
 
-  // TODO: fetch instance and zone from metadata. This will require function to
-  // be asynchrous.
-  if (normalizedConfig.instance === undefined) {
-    normalizedConfig.instance = '';
-  }
-  if (normalizedConfig.zone === undefined) {
-    normalizedConfig.zone = '';
+  if (normalizedConfig.serviceContext.service === undefined) {
+    throw new Error('service must be specified in the configuration');
```

Comment:
Is having the error message start with lowercase and not have a period at the end intentional? I just don't know what's the best style for these but make sure it's consistent with other places of reporting an error.

Reply:
I looked at trace -- seems like capitalization with period is used there. 
I've fix this here.